pkg/setup/node_test.go

```diff
@@ -17,6 +20,215 @@ import (
+			for pk, r := range routers {
+				mr := r.(*mockRouterGateway)
+				t.Logf("Checking router %s: lastRtID=%d edgeRules=%d interRules=%d",
+					pk, mr.lastRtID, len(mr.edgeRules), len(mr.interRules))
+				checkRtIDKeysOfRouterRules(t, mr)
+			}
+
+			// TODO: assert: edge routers
+			// * Ensure edge routers have 1 edge rule each, and no inter rules.
+			// * Edge rule's descriptor should be of provided src/dst pk/port.
+
+			// TODO: assert: inter routers
+			// * Ensure inter routers have 2 or more inter rules (depending on routes).
+			// * Ensure inter routers have no edge rules.
+		})
+	}
+}
+
+// checkRtIDKeysOfRouterRules ensures that the rules advertised to the router (from the setup logic) has route ID keys
+// which are valid.
+func checkRtIDKeysOfRouterRules(t *testing.T, r *mockRouterGateway) {
+	r.mx.Lock()
+	defer r.mx.Unlock()
+
+	var rtIDKeys []routing.RouteID
+
+	for _, edge := range r.edgeRules {
+		rtIDKeys = append(rtIDKeys, edge.Forward.KeyRouteID(), edge.Reverse.KeyRouteID())
+	}
+	for _, rules := range r.interRules {
+		for _, rule := range rules {
+			rtIDKeys = append(rtIDKeys, rule.KeyRouteID())
+		}
+	}
+
+	// assert: no duplicate rtIDs
+	dupM := make(map[routing.RouteID]struct{})
+	for _, rtID := range rtIDKeys {
+		dupM[rtID] = struct{}{}
+	}
+	assert.Len(t, dupM, len(rtIDKeys), "rtIDKeys=%v dupM=%v", rtIDKeys, dupM)
+
+	// assert: all routes IDs are explicitly reserved by router
+	for _, rtID := range rtIDKeys {
+		assert.LessOrEqual(t, uint32(rtID), r.lastRtID)
+	}
+}
+
+func countUniquePKs(pks []cipher.PubKey) int {
+	m := make(map[cipher.PubKey]struct{})
+	for _, pk := range pks {
+		m[pk] = struct{}{}
+	}
+	return len(m)
+}
+
+func biRouteFromKeys(fwdPKs, revPKs []cipher.PubKey, srcPort, dstPort routing.Port) routing.BidirectionalRoute {
+	fwdHops := make([]routing.Hop, len(fwdPKs)-1)
+	for i, srcPK := range fwdPKs[:len(fwdPKs)-1] {
+		dstPK := fwdPKs[i+1]
+		fwdHops[i] = routing.Hop{TpID: determineTpID(srcPK, dstPK), From: srcPK, To: dstPK}
+	}
+	revHops := make([]routing.Hop, len(revPKs)-1)
+	for i, srcPK := range revPKs[:len(revPKs)-1] {
+		dstPK := revPKs[i+1]
+		revHops[i] = routing.Hop{TpID: determineTpID(srcPK, dstPK), From: srcPK, To: dstPK}
+	}
+	// TODO: This should also return a map of format: map[uuid.UUID][]cipher.PubKey
+	// This way, we can associate transport IDs to the two transport edges, allowing for more checks.
+	return routing.BidirectionalRoute{
+		Desc:      routing.NewRouteDescriptor(fwdPKs[0], revPKs[0], srcPort, dstPort),
+		KeepAlive: 0,
+		Forward:   fwdHops,
+		Reverse:   revHops,
+	}
+}
+
+// for tests, we make transport IDs deterministic
+// hence, we can derive the tpID from any pk pair
+func determineTpID(pk1, pk2 cipher.PubKey) (tpID uuid.UUID) {
+	v1, v2 := pk1.Big(), pk2.Big()
+	var hash cipher.SHA256
+	if v1.Cmp(v2) > 0 {
+		hash = cipher.SumSHA256(append(pk1[:], pk2[:]...))
+	} else {
+		hash = cipher.SumSHA256(append(pk2[:], pk1[:]...))
+	}
```

Comment:
Could you please add empty lines before and after `if` block? I think it may improve readability a bit.